x-pack/plugin/transform/src/main/java/org/elasticsearch/xpack/transform/transforms/pivot/Pivot.java

```diff
@@ -147,33 +205,58 @@ public SearchRequest buildSearchRequest(SourceConfig sourceConfig, Map<String, O
         return searchRequest;
     }
 
-    public AggregationBuilder buildAggregation(Map<String, Object> position, int pageSize) {
+    @Override
+    public SearchSourceBuilder buildSearchQuery(SearchSourceBuilder builder, Map<String, Object> position, int pageSize) {
         cachedCompositeAggregation.aggregateAfter(position);
         cachedCompositeAggregation.size(pageSize);
 
-        return cachedCompositeAggregation;
+        return builder.size(0).aggregation(cachedCompositeAggregation);
     }
 
-    public CompositeAggregationBuilder buildIncrementalBucketUpdateAggregation(int pageSize) {
+    @Override
+    public ChangeCollector buildChangeCollector(String synchronizationField) {
+        Map<String, FieldCollector> fieldCollectors = new HashMap<>();
 
-        CompositeAggregationBuilder compositeAgg = createCompositeAggregationSources(config, true);
-        compositeAgg.size(pageSize);
-
-        return compositeAgg;
-    }
-
-    public Map<String, Set<String>> initialIncrementalBucketUpdateMap() {
-
-        Map<String, Set<String>> changedBuckets = new HashMap<>();
         for (Entry<String, SingleGroupSource> entry : config.getGroupConfig().getGroups().entrySet()) {
-            if (entry.getValue().supportsIncrementalBucketUpdate()) {
-                changedBuckets.put(entry.getKey(), new HashSet<>());
+            switch (entry.getValue().getType()) {
```

Comment:
It seems here we are effectively switching on a class type. Why can't we take advantage of polymorphism?

Reply:
agree, this part disturbs me, too. Its error-prone, when adding a new group source, this can be easily forgotten and you won't realize until you use it in continuous mode.